apps/homescreen/js/group.js

```diff
@@ -0,0 +1,305 @@
+     * 4 Set an offset and size on the background and container so they still
+     *   remain in the same position.
+     * 5 Set transform on background so it expands into the full group area.
+     * 6 Set style class on the icons container fades out.
+     * 7 After the icon container fades out, change its style properties to
+     *   fill the expanded group.
+     * 8 Synchronise icon container.
+     * 9 Set style property on icon container to fade it in.
+     */
+    this.state = EXPANDING;
+
+    // Part 1.
+    document.body.classList.add('expanding');
+
+    // Part 2.
+    var rect = this.getBoundingClientRect();
+    var originalWidth = this.clientWidth;
+    var originalHeight = this.clientHeight;
+    var originalLeft = this.clientLeft;
+    var originalTop = this.clientTop;
+
+    var parentOffsetTop = parent.offsetTop;
+    var targetLeft = Math.round(-rect.left);
+    var targetTop = Math.round(parentOffsetTop - rect.top);
+    var targetWidth = window.innerWidth;
+    var targetHeight = window.innerHeight - parentOffsetTop;
+
+    // Use absolute positioning instead of a transform so the fixed-positioning
+    // used when dragging works correctly.
+    parent.setUseTransform(this.parentNode, false);
+
+    // We need opened groups to appear above the app grid. We can only
+    // do this by setting a z-index on the gaia-container-child due to
+    // the established stacking order.
+    this.parentNode.parentNode.style.zIndex = '1';
+
+    // Shrink the target rect depending on SIZE_RATIO
+    if (targetHeight >= targetWidth) {
+      targetTop += (targetHeight * (1 - SIZE_RATIO)) / 2;
+      targetHeight *= SIZE_RATIO;
+    } else {
+      targetLeft += (targetWidth * (1 - SIZE_RATIO)) / 2;
+      targetWidth *= SIZE_RATIO;
+    }
+
+    // Part 3.
+    this.style.left = targetLeft + 'px';
+    this.style.top = targetTop + 'px';
+    this.style.width = targetWidth + 'px';
+    this.style.height = targetHeight + 'px';
+
+    // Part 4.
+    this.container.style.left =
+      this.background.style.left = (originalLeft - targetLeft) + 'px';
+    this.container.style.top =
+      this.background.style.top = (originalTop - targetTop) + 'px';
+    this.container.style.width =
+      this.background.style.width = originalWidth + 'px';
+    this.container.style.height =
+      this.background.style.height = originalHeight + 'px';
+
+    // Part 5.
+    var bgOffsetLeft = Math.round((targetWidth / 2) - (originalWidth / 2));
+    var bgOffsetTop = Math.round((targetHeight / 2) - (originalHeight / 2));
+    var bgLeft = Math.round(bgOffsetLeft + targetLeft) - originalLeft;
+    var bgTop = Math.round(bgOffsetTop + targetTop) - originalTop;
+    var bgTargetSize =
+      Math.sqrt(2 * Math.pow(Math.max(targetWidth, targetHeight), 2));
+    var bgScale = bgTargetSize / Math.max(originalWidth, originalHeight);
+    this.background.style.transform =
+      'translate(' + bgLeft + 'px, ' + bgTop + 'px) scale(' + bgScale + ')';
```

Comment:
Typically this kind of string is usually more readable when formatted with back tick `.